codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +321,509 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+        }
+
+        @Override
+        protected void doDisconnect() throws Exception {
+            doClose();
+        }
+
+        @Override
+        protected void doClose() throws Exception {
+            closed = true;
+
+            if (!streamClosedWithoutError && isStreamIdValid(stream().id())) {
+                Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).stream(stream());
+                ctx.writeAndFlush(resetFrame);
+            }
+
+            while (!inboundBuffer.isEmpty()) {
+                ReferenceCountUtil.release(inboundBuffer.poll());
             }
         }
-        return future;
-    }
 
-    @SuppressWarnings("unchecked")
-    private static void initOpts(Channel channel, Map<ChannelOption<?>, Object> opts) {
-        if (opts != null) {
-            for (Entry<ChannelOption<?>, Object> e: opts.entrySet()) {
-                try {
-                    if (!channel.config().setOption((ChannelOption<Object>) e.getKey(), e.getValue())) {
-                        logger.warn("Unknown channel option: " + e);
+        @Override
+        protected void doBeginRead() {
+            if (readInProgress) {
+                return;
+            }
+
+            final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+            allocHandle.reset(config());
+            if (inboundBuffer.isEmpty()) {
+                readInProgress = true;
+                return;
+            }
+
+            do {
+                Object m = inboundBuffer.poll();
+                if (m == null) {
+                    break;
+                }
+                if (!doRead0(m, allocHandle)) {
+                    // Channel closed, and already cleaned up.
+                    return;
+                }
+            } while (allocHandle.continueReading());
+
+            allocHandle.readComplete();
+            pipeline().fireChannelReadComplete();
+        }
+
+        @Override
+        protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+            if (closed) {
+                throw CLOSED_CHANNEL_EXCEPTION;
+            }
+
+            try {
+                if (sizeEstimatorHandle == null) {
+                    sizeEstimatorHandle = config().getMessageSizeEstimator().newHandle();
+                }
+                for (;;) {
+                    final Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    if (msg == REEVALUATE_WRITABILITY_MESSAGE) {
+                        in.remove();
+                        continue;
+                    }
+
+                    final int bytes = sizeEstimatorHandle.size(msg);
+
+                    /*
+                     * The flow control window needs to be decrement before stealing the message from the buffer (and
+                     * thereby decrementing the number of pending bytes). Else, when calling steal() the number of
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)